src/js/client/editprefset-dialogs-widget.js

```diff
@@ -0,0 +1,304 @@
+        path: oldPath,
+        type: "DELETE"
+    });
+
+    transaction.fireChangeRequest({
+        path: path,
+        value: prefsetToEdit
+    });
+
+    transaction.commit();
+    that.closeDialog();
+};
+
+
+/**
+ * Confirm deletion of NP Set Context Dialog
+ */
+fluid.defaults("gpii.devpmt.dialogs.confirmDeleteContextDialog", {
+    gradeNames: ["gpii.devpmt.dialogs.confirmDialog"],
+    templates: {
+        initial: "editprefset-confirmDeleteContext-dialog"
+    },
+    model: {
+        contextId: "" // Should be populated/relayed during construction before showing dialog
+    },
+    invokers: {
+        acceptConfirmDialog: {
+            funcName: "gpii.devpmt.dialogs.confirmDeleteContextDialog.acceptConfirmDialog",
+            args: ["{that}", "{gpii.devpmt.editPrefs}", "{that}.model.contextId"]
+        }
+    }
+});
+
+gpii.devpmt.dialogs.confirmDeleteContextDialog.acceptConfirmDialog = function (that, editPrefs, contextId) {
+    that.closeDialog();
+    var path = "flatPrefs.contexts." + contextId;
+    editPrefs.applier.change(path, false, "DELETE");
+};
+
+/**
+ * Add Product to NP Set Dialog
+ */
+fluid.defaults("gpii.devpmt.dialogs.confirmAddProductDialog", {
+    gradeNames: ["gpii.devpmt.dialogs.confirmDialog"],
+    templates: {
+        initial: "editprefset-confirmAddProduct-dialog"
+    },
+    model: {
+        appId: null,
+        name: null // Human Name of Product
+    },
+    invokers: {
+        acceptConfirmDialog: {
+            funcName: "gpii.devpmt.dialogs.confirmAddProductDialog.acceptConfirmDialog",
+            args: ["{that}", "{that}.model.appId", "{gpii.devpmt.editPrefs}.model.contextNames", "{gpii.devpmt.editPrefs}.editProductEnabled"]
+        }
+    }
+});
+
+/**
+ * Clicking Add on the Product Dialog adds this product to the preference set.
+ * A preference safe must have at least one context (prefset) in order to add
+ * a product to it.
+ *
+ * @param  {Object} that               Dialog instance
+ * @param  {String} appId              String indicating the appId
+ *                                     ex. http://registry.gpii.net/applications/com.android.freespeech
+ * @param  {Array} contextNames        Array of Strings with the name/id the context is
+ *                                     keyed by. The product will be initially added to the first
+ *                                     item in this array.
+ * @param  {Function} editProductEnabled Invoker from `gpii.devpmt.editPrefs` to add the
+ *                                       product to.
+ */
+gpii.devpmt.dialogs.confirmAddProductDialog.acceptConfirmDialog = function (that, appId, contextNames, editProductEnabled) {
+    // In this case we actually need to close the dialog first... as the page
+    // rerenders based on a model listener when the product is enabled, and does
+    // wonky things... such as removing the ability to vertical scroll. Should look
+    // at reworking this perhaps.
+    that.closeDialog();
```

Comment:
Again, shall the screening of input values be performed first?

Reply:
I added a check to ensure the appId is in the solutions registry.